libs/vm/src/module.cpp

```diff
@@ -143,72 +149,122 @@ double toFloat64(VM * /* vm */, AnyPrimitive const &from)
+{
+  interfaces.address.CreateConstuctor<>();
+  interfaces.address.CreateConstuctor<Ptr<String>>();
+
+  interfaces.iarray.CreateConstuctor<int32_t>();
+
+  interfaces.imap.CreateConstuctor<>();
+
+  interfaces.imatrix.CreateConstuctor<int32_t, int32_t>();
+
+  interfaces.istate.CreateConstuctor<Ptr<String>, TemplateParameter>();
+  interfaces.istate.CreateConstuctor<Ptr<Address>, TemplateParameter>();
+}
+
+void EnableIndexOperators(AllClassInterfaces &interfaces)
+{
+  interfaces.iarray.EnableIndexOperator<AnyInteger, TemplateParameter>();
+  interfaces.imap.EnableIndexOperator<TemplateParameter1, TemplateParameter2>();
+  interfaces.imatrix.EnableIndexOperator<AnyInteger, AnyInteger, TemplateParameter>();
+}
+
+void CreateMemberFunctions(AllClassInterfaces &interfaces)
+{
+  interfaces.address.CreateMemberFunction("signedTx", &Address::HasSignedTx);
+
+  interfaces.iarray.CreateMemberFunction("append", &IArray::Append);
+  interfaces.iarray.CreateMemberFunction("count", &IArray::Count);
+  interfaces.iarray.CreateMemberFunction("extend", &IArray::Extend);
+  interfaces.iarray.CreateMemberFunction("popBack", &IArray::PopBackOne);
+  interfaces.iarray.CreateMemberFunction("popBack", &IArray::PopBackMany);
+  interfaces.iarray.CreateMemberFunction("popFront", &IArray::PopFrontOne);
+  interfaces.iarray.CreateMemberFunction("popFront", &IArray::PopFrontMany);
+  interfaces.iarray.CreateMemberFunction("reverse", &IArray::Reverse);
+
+  interfaces.imap.CreateMemberFunction("count", &IMap::Count);
+
+  interfaces.istate.CreateMemberFunction("existed", &IState::Existed);
+  interfaces.istate.CreateMemberFunction("get", &IState::Get);
+  interfaces.istate.CreateMemberFunction("set", &IState::Set);
+
+  interfaces.string.CreateMemberFunction("find", &String::Find);
+  interfaces.string.CreateMemberFunction("length", &String::Length);
+  interfaces.string.CreateMemberFunction("reverse", &String::Reverse);
+  interfaces.string.CreateMemberFunction("substr", &String::Substring);
+  interfaces.string.CreateMemberFunction("trim", &String::Trim);
+}
+
+void CreateFreeFunctions(Module *m)
+{
+  m->CreateFreeFunction("toInt8", &toInt8);
+  m->CreateFreeFunction("toByte", &toByte);
+  m->CreateFreeFunction("toInt16", &toInt16);
+  m->CreateFreeFunction("toUInt16", &toUInt16);
+  m->CreateFreeFunction("toInt32", &toInt32);
+  m->CreateFreeFunction("toUInt32", &toUInt32);
+  m->CreateFreeFunction("toInt64", &toInt64);
+  m->CreateFreeFunction("toUInt64", &toUInt64);
+  m->CreateFreeFunction("toFloat32", &toFloat32);
+  m->CreateFreeFunction("toFloat64", &toFloat64);
+}
+
+AllClassInterfaces GetAllClassInterfaces(Module *m)
+{
+  // Ordering must match that in AllClassInterfaces
+  return {m->GetClassInterface<Address>(), m->GetClassInterface<IArray>(),
+          m->GetClassInterface<IMap>(),    m->GetClassInterface<IMatrix>(),
+          m->GetClassInterface<IState>(),  m->GetClassInterface<String>()};
+}
+
+}  // namespace
+
 Module::Module()
```

Comment:
I would suggest to go back to the original concept (put everything back in to this methob), but use the builder pattern as it was shown in above comment https://github.com/fetchai/ledger/pull/1142/files#r290517903.

Reply:
See my answer above.

Reply:
I think we stay with the original code, but change it to use the superior Builder pattern in general, as Peter suggests, but deal with interdependency by having multiple builder chains where necessary. The present difficulty (String depending on Array < String > ) can be solved by simply moving a GetClassInterface < String >()... chain to the end.

Reply:
In the general cyclic case, A depends on B, B depends on A, you can create type A, create type B and _then_ define their member functions etc in two subsequent ClassInterface chains.

Reply:
@robertdickson that's my worry here: eventually in order to account for these interdependencies, we'll have to set up bits of interfaces A and B (instantiations), followed by the rest of A and B (members etc.), and in the general case the `Module()` ctor becomes a big blob of stuff, all in one function. My rationale here was that we can get ahead of that mess and segregate the different parts of the interfaces' setup by category (first create all instantiations of all interfaces, then all members, etc.) rather than by interface (first we setup everything related to String, then everything related to Array...). In the former setup, here we don't have to fiddle with reordering lines within each category, ever again, because the categories are not interdependent, while the interfaces in general are, because instantiations need to be registered before they are used.

Another way to tackle this would be to make function and method definitions lazily validate that all the types had been registered with the VM after Module is constructed, but that's a big gun to roll out for such a small problem.

I'll reorder the String and Array block to make `split()` work and introduce the builder pattern, but I think that instantiating all the interfaces first would be the sustainable way to go.

Reply:
See my comment https://github.com/fetchai/ledger/pull/1142#discussion_r290641730 above.

Reply:
This a solution in search of a problem at the moment....let's move on.